### PR TITLE
HBW-398 Add services select table

### PR DESCRIPTION
**Pull request by EternalStuden1**
services_table

### CODE REVIEW

File: hbw/app/models/hbw/fields/services_table.rb [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+
+      def customer_id
+        get_bpm_prop_value('customer_id_variable')
+      end
+
+      def contract_id
+        get_bpm_prop_value('contract_id_variable')
+      end
+
+      def currency_code
```
Comment (TimAle) [inline hbw/app/models/hbw/fields/services_table.rb:271]: we will do it after introducing a lib for money formatting

File: hbw/app/javascript/packs/hbw/components/form/services_table.tsx [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+  vlan?: number,
+  phone?: string
+  rowState: number
+}
+
+type Service = TableRow & {
+  childServices: TableRow[]
+}
+
+type HeaderEntity = {
```
[the remark this answers is not in the document]
Comment (TimAle) [reply hbw/app/javascript/packs/hbw/components/form/services_table.tsx:79]: or even `Columns`

**Review by TimAle — CHANGES_REQUESTED** [review]

File: hbw/app/javascript/packs/hbw/components/form/services_table.tsx [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+
+function getVisibleHeaders (tableHeaders: HeaderEntity[], hiddenHeaders: string[]) {
+  return [...tableHeaders].filter(header => !hiddenHeaders.includes(header.accessor));
+}
+
+// eslint-disable-next-line max-len
+function getHeadersToShow (tableHeaders: HeaderEntity[], hiddenHeaders: string[] | undefined, defaultHiddenHeaders: string[]) {
+  if (hiddenHeaders !== undefined) {
+    return getVisibleHeaders(tableHeaders, hiddenHeaders);
+  }
```
[the remark this answers is not in the document]
Comment (EternalStuden1) [reply hbw/app/javascript/packs/hbw/components/form/services_table.tsx:100]: There are 2 different options: hiddenHeaders form definition is set to [] and not set at all. In the first case we should show all headers, in the second case - hide default list of headers. That's why it's important to have undefined | string[]

File: hbw/app/javascript/packs/hbw/components/form/services_table.tsx [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+  equipmentName: string,
+  servicePrice: number,
+  serviceQuantity?: number,
+  serviceUnit: string,
+  serviceTotalPrice: number,
+  subscriptionId: number,
+  parSubscriptionId?: number,
+  subscriptionBeginDate: string,
+  subscriptionEndDate?: string,
+  allowProvisioning: number,
```
[the remark this answers is not in the document]
Comment (EternalStuden1) [reply hbw/app/javascript/packs/hbw/components/form/services_table.tsx:66]: Value from db is fetched as constant (`CL_Creating_Allowed`/`CL_Creating_Forbidden`), thus it's type is number

File: hbw/app/javascript/packs/hbw/constants/c_l_creating_state.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+// eslint-disable-next-line no-shadow
```
[the remark this answers is not in the document]
Comment (EternalStuden1) [reply hbw/app/javascript/packs/hbw/constants/c_l_creating_state.ts:1]: Снимок экрана 2023-10-11 в 12 48 31

**Review by TimAle — CHANGES_REQUESTED** [review]

File: hbw/app/javascript/packs/hbw/components/form/services_table.tsx [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+      { isColumnVisible(columns, 'macAddress') && <td className="white-space-no-wrap">
+        {serviceRow.macAddress}
+      </td> }
+      { isColumnVisible(columns, 'ipAddress') && <td className="white-space-no-wrap">
+        {serviceRow.ipAddress}
+      </td> }
+      { isColumnVisible(columns, 'vlan') && <td className="white-space-no-wrap">
+        {serviceRow.vlan}
+      </td> }
+      { isColumnVisible(columns, 'phone') && <td className="white-space-no-wrap">
```
[the remark this answers is not in the document]
Comment (EternalStuden1) [reply hbw/app/javascript/packs/hbw/components/form/services_table.tsx:232]: No. Column with service address should not be whitespaced as its content wrap to the next line is acceptable

[the remark this answers is not in the document]
Comment (TimAle) [reply hbw/app/javascript/packs/hbw/components/form/services_table.tsx:232]: but...
image

File: hbw/app/models/hbw/fields/services_table.rb [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+
+      def get_db_value
+        sql = get_db_value_sql % {
+          account_id:      account_id,
+          contract_id:     contract_id,
+          equipment_types: equipment.map { |equipment_entry| equipment_entry.fetch('id').to_i }.join(',')
+        }
+        db_value = loader(sql, {}).load
+        db_value.map do |subscription|
+          subscription['allowProvisioning'] = subscription['allowProvisioning'] == 1212
```
Comment (TimAle) [inline hbw/app/models/hbw/fields/services_table.rb:146]: magic number

**Review by TimAle — CHANGES_REQUESTED** [review]

File: hbw/app/models/hbw/fields/services_table.rb [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+      "
+
+      def get_db_value
+        sql = get_db_value_sql % {
+          account_id:,
+          contract_id:,
+          equipment_types: equipment.map { |equipment_entry| equipment_entry.fetch('id').to_i }.join(',')
+        }
+        db_value = loader(sql, {}).load
+        binding.pry
```
Comment (TimAle) [inline hbw/app/models/hbw/fields/services_table.rb:147]:
```suggestion

```

**Review by TimAle — APPROVED** [review]

**Pull request merged**